network-libp2p/src/network.rs

```diff
@@ -1298,6 +1352,149 @@ impl Network {
     pub fn metrics(&self) -> Arc<NetworkMetrics> {
         self.metrics.clone()
     }
+
+    fn is_under_the_rate_limits<Req: RequestCommon>(
+        peer_request_limits: Arc<Mutex<HashMap<PeerId, HashMap<u16, RateLimit>>>>,
+        peer_id: PeerId,
+        request_id: RequestId,
+    ) -> bool {
+        // Gets lock of peer requests limits read and write on it.
+        let mut peer_request_limits = peer_request_limits.lock();
+
+        // If the peer has never sent a request of this type, creates a new entry.
+        let requests_limit = peer_request_limits
+            .entry(peer_id)
+            .or_default()
+            .entry(Req::TYPE_ID)
+            .or_insert_with(|| {
+                RateLimit::new(Req::MAX_REQUESTS, Req::TIME_WINDOW, TokioInstant::now())
+            });
+
+        // Ensures that the request is allowed based on the set limits and updates the counter.
+        // Returns early if not allowed.
+        if !requests_limit.increment_and_is_allowed(1, TokioInstant::now()) {
+            log::debug!(
+                "[{:?}][{:?}] {:?} Exceeded max requests rate {:?} requests per {:?} seconds",
+                request_id,
+                peer_id,
+                std::any::type_name::<Req>(),
+                Req::MAX_REQUESTS,
+                Req::TIME_WINDOW,
+            );
+            return false;
+        }
+        true
+    }
+
+    fn remove_rate_limits(
+        peer_request_limits: Arc<Mutex<HashMap<PeerId, HashMap<u16, RateLimit>>>>,
+        rate_limits_pending_deletion: Arc<Mutex<VecDeque<((PeerId, u16), TokioInstant)>>>,
+        peer_id: PeerId,
+    ) {
+        // Every time a peer disconnects, we delete all expired pending limits.
+        Self::clean_up(
+            Arc::clone(&peer_request_limits),
+            Arc::clone(&rate_limits_pending_deletion),
+        );
+
+        // Firstly we must acquire the lock of the pending deletes to avoid deadlocks.
+        let mut rate_limits_pending_deletion_l = rate_limits_pending_deletion.lock();
+        let mut peer_request_limits_l = peer_request_limits.lock();
+
+        // Go through all existing request types of the given peer and deletes the limit counters if possible or marks it for deletion.
+        if let Some(request_limits) = peer_request_limits_l.get_mut(&peer_id) {
+            request_limits.retain(|req_type, rate_limit| {
+                // Gets the requests limit and deletes it if no counter info would be lost, otherwise places it as pending deletion.
+                if !rate_limit.can_delete(TokioInstant::now()) {
+                    rate_limits_pending_deletion_l
+                        .push_back(((peer_id, *req_type), rate_limit.next_reset_time()));
```

Comment:
If I'm not mistaken, this could lead to adding the same thing multiple times, does it?

Reply:
And if so, is it a problem?

Reply:
If we receive two connection closed messages we can try to remove the rate limits of the said peer again and place the limits again on the queue for pending deletion queue. Those will be cleared out similarly at the same time when the expiration date has passed. 

So duplicates are possible in case of duplicate connection closed messages but they would be cleared out all at the same time eventually (max time_window defined).

Reply:
That sounds good to me then.